feat: In trace sampler

blumamir opened the pull request:
~**Still looking for a sponsor for this component 🙏**~ Sponsor found!

**Description:** Add a new processor for sampling out individual spans from a complete trace.

**Link to tracking Issue:** #17086 

**Testing:** added tests similar to how I saw in other processors. I also ran the collector with the processor in an actual setup with real instrumentations and observed it was functionalizing correctly.

**Documentation:** added a README file similar to other processors.

I am not native to golang, and this is my first contribution to this repo. Please review thoroughly and feel free to share how I can improve and contribute high-quality code. Thank you!

Commit:
feat: in-trace sampler processor

Commit:
Merge remote-tracking branch 'upstream/main' into in-trace-sampler

Commit:
add chloggen entry

Commit:
chore: addlicense

Commit:
fix: typo

Commit:
chore: fix most of the lints

atoulme reviewed (commented):
Good start, found a few typos and small fixes.

atoulme commented:
@blumamir are you still interested in working on this?

blumamir commented:
> @blumamir are you still interested in working on this?

Hi @atoulme , Thank you for the review, and sorry for the delay, I missed the notifications. 
I am still interested. Going over your comments now 🙏

blumamir committed:
Update processor/intracesampler/README.md

Co-authored-by: Antoine Toulme <antoine@toulme.name>

blumamir committed:
Update processor/intracesampler/README.md

Co-authored-by: Antoine Toulme <antoine@toulme.name>

blumamir committed:
Update processor/intracesampler/processor.go

Co-authored-by: Antoine Toulme <antoine@toulme.name>

blumamir committed:
Update processor/intracesampler/processor.go

Co-authored-by: Antoine Toulme <antoine@toulme.name>

blumamir committed:
Update processor/intracesampler/processor.go

Co-authored-by: Antoine Toulme <antoine@toulme.name>

Commit:
fix: don't panic on empty trace batch

blumamir commented on processor/intracesampler/processor.go:
```diff
@@ -0,0 +1,218 @@
+// this indicates that the processor is run after another processor
+// that emits completed traces after timeout
+func isAllSameTraceID(td ptrace.Traces) bool {
+	firstTraceID := td.ResourceSpans().At(0).ScopeSpans().At(0).Spans().At(0).TraceID()
```
Right, good catch.
Refactored the code + added unit test for that

blumamir commented on processor/intracesampler/murmur3.go:
```diff
@@ -0,0 +1,69 @@
+package intracesampler
+
+// hash is a murmur3 hash function, see http://en.wikipedia.org/wiki/MurmurHash
+func hash(key []byte, seed uint32) (hash uint32) {
```
fair enough.
Can you please suggest the best way to do it in go? should I add a new module with the murmur3 code and consume it from both probabilistic sampler and this sampler?

atoulme replied:
You would move the code to pkg/internal so both components can import it. It can be dine with a follow up PR, just file an issue for it.

blumamir replied:
looks like the murmur3 hash was replaced with another hash [here](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/17837) after I opened the PR.
I will update the code accordingly

blumamir replied:
Updated the PR with the new hash algorithm, CI is green. I think it will be best to extract the common code after this one merge so this PR will not make changes to other components.

Should I open the issue now or after we merge?

blumamir commented:
> Good start, found a few typos and small fixes.

Thanks @atoulme , addressed all your comments. I need help understanding what is the idiomatic way to share the murmur3 code between different modules. Do you know why the murmur3 implementation was copied to this codebase instead of using a dedicated module like [this one](https://pkg.go.dev/github.com/spaolacci/murmur3)?

Commit:
Merge remote-tracking branch 'upstream/main' into in-trace-sampler

Commit:
chore: add codeowner

Commit:
chore: upgrade colletor modules dependencies

Commit:
chore: lint

Commit:
fix: lint issues

Commit:
chore: crosslink

Commit:
chore: add new processor to versions.yaml

Commit:
fix: name of processor in versions.yaml

Commit:
fix: versions.yaml

Commit:
chore: update github templates

blumamir commented:
I fixed the CI issues and now it is green.
Please let me know if there is anything still missing and how to progress with the PR.

Thanks again!

atoulme commented:
I don’t know why murmur3 was inlined rather than imported. Maybe a thing to consider.

Commit:
feat: use hnv hash instead of murmur3

Commit:
Merge remote-tracking branch 'upstream/main' into in-trace-sampler

Commit:
chore: make goporto

blumamir commented:
@atoulme - anything else left to do before it can merge?

atoulme reviewed (approved):
LGTM

MovieStoreGuy commented on processor/intracesampler/config.go:
```diff
@@ -0,0 +1,43 @@
+	HashSeed uint32 `mapstructure:"hash_seed"`
+
+	// will unsample spans from this scope, if they have no sampled descendants
+	ScopeLeaves []string `mapstructure:"scope_leaves"`
```
`ScopeLeaves` doesn't feel like the right name here, this should use an existing naming with Open Telemetry.

MovieStoreGuy replied:
I believe that is more instrumentation scope?  As for a better name, I haven't yet figured out what would make the most sense.

MovieStoreGuy commented on processor/intracesampler/config_test.go:
```diff
@@ -0,0 +1,60 @@
+)
+
+func TestLoadConfig(t *testing.T) {
+	cm, err := confmaptest.LoadConf(filepath.Join("testdata", "config.yaml"))
```
If you can, please add `t.Parallel()` with tests so that these can run concurrently.

MovieStoreGuy commented on processor/intracesampler/fnvhasher.go:
```diff
@@ -0,0 +1,37 @@
+)
+
+// computeHash creates a hash using the FNV-1a algorithm
+func computeHash(b []byte, seed []byte) uint32 {
```
Is there any reason to use a uint32 instead of a uint64?

blumamir replied:
I took this code from the existing implementation in [probabilistic sampler processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/85198a60c315d75fcf74b6b84a6504a6e699e00a/processor/probabilisticsamplerprocessor/fnvhasher.go#L22).
My goal was to apply the same hash computation so they can be chained in the same pipeline and sample consistently.

I suggest addressing this in a followup PR which will also align the behavior for both processor by extracting the function into a common package.

MovieStoreGuy replied:
This makes sense to me.

MovieStoreGuy commented on processor/intracesampler/fnvhasher.go:
```diff
@@ -0,0 +1,37 @@
+
+// i32tob converts a seed to a byte array to be used as part of fnv.Write()
+func i32tob(val uint32) []byte {
+	r := make([]byte, 4)
```
Make sure you define magic numbers somewhere in case anyone needs to change at least understand what they are intended for.

blumamir replied:
This too is copied as-is from the existing [probabilistic sampler processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/85198a60c315d75fcf74b6b84a6504a6e699e00a/processor/probabilisticsamplerprocessor/fnvhasher.go#L22). 

I prefer if this improvement be done in a dedicated PR for both places.

MovieStoreGuy replied:
That is also fair, if you can use a perma link as to where this code is referenced, it would help those wanting to figure out the reasoning for this code and how it came to existence. ( Bitshifting is a pain to get right the first time in my experience.)

blumamir replied:
added relevant comment in code

MovieStoreGuy commented on processor/intracesampler/go.mod:
```diff
@@ -0,0 +1,77 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/processor/intracesamplerprocessor
+
+go 1.18
```
Pleas set this to `go 1.19` to be aligned with project's root `go.mod`.

MovieStoreGuy commented on processor/intracesampler/processor.go:
```diff
@@ -0,0 +1,224 @@
+		processorhelper.WithCapabilities(consumer.Capabilities{MutatesData: true}))
+}
+
+type FullSpan struct {
```
Please make this type private.

MovieStoreGuy commented on processor/intracesampler/processor.go:
```diff
@@ -0,0 +1,224 @@
+	span     ptrace.Span
+}
+
+type TraceTreeData struct {
```
Please make this type private.

MovieStoreGuy commented on processor/intracesampler/processor.go:
```diff
@@ -0,0 +1,224 @@
+			for k := 0; k < spans.Len(); k++ {
+				span := spans.At(k)
+
+				fullspan := FullSpan{
+					resource: resource,
+					scope:    scope,
+					span:     span,
+				}
```
In order to reduce on the number of required allocation, you can make this:

```golang
fullspan := FullSpan{
    resource : rs.Resource(),
    scope     : ss.Scope(),
    span       : span, 
}
```

blumamir replied:
no problem, but isn't it cheaper to cache the getter function result once and use it for all iterations of the loop? instead of invoking it again and again for the nested loops?

MovieStoreGuy commented on processor/intracesampler/processor.go:
```diff
@@ -0,0 +1,224 @@
+		}
+	}
+
+	// find roots
```
I'd have to read the spec, but I vaguely remember that for a span to be a root, the SpanID and TraceID will be equal?

MovieStoreGuy replied:
I was wrong on this, there is no reference of the traceid and span id being equal if it is a root span. Not sure where I read this.

MovieStoreGuy commented on processor/intracesampler/processor.go:
```diff
@@ -0,0 +1,224 @@
+// that emits completed traces after timeout
+// if a single trace id is found, it is returend, otherwise nil is returned
+func getSingleTraceID(td ptrace.Traces) *pcommon.TraceID {
+	var traceID *pcommon.TraceID
```
Please avoid using using pointers to `pdata` types since it impacts garbage collection.

blumamir replied:
No problem, but just so I understand what I am doing and applying a correct fix - is it a problem specifically with `pdata` or is it bad practice to use pointers in general?

Can you please point me to some resources that explain what is the idiomatic garbage collection-friendly alternative?

MovieStoreGuy replied:
Only for pdata types in the general case, the reason being is that pdata is an optimised layer above the raw proto types that are allocated on the heap. 

So potentially having the reference type also on the heap does mean there is some performance impacted on look up, (The pdata wrappers themselves are tiny so they aren't an issue on the stack) but it also makes it harder for the GC to know when to cleanup the types due to the indirections going on.

blumamir replied:
I removed the reference as requested.
Please note that now to indicate "empty" trace id, I use `pcommon.NewTraceIDEmpty()`. I think it means that instead of comparing a pointer to nil, we now have to compare two `[16]byte` arrays to each other in each iteration which I guess also comes with a performance impact

MovieStoreGuy commented on processor/intracesampler/processor.go:
```diff
@@ -0,0 +1,224 @@
+
+	// currrent span should be unsampled if it's in the unsampledScopes map
+	// and all its children are also unsampled.
+	currentUnsampled := slices.Contains(its.config.ScopeLeaves, currentScopeName)
```
I feel like you would be better of using map like a set instead of `slice.Contains`.

You can use:
```golang
map[pcommon.SpanID]struct{}
```

This will allocate no memory for the empty struct and will be more optimal than `slices.Contains`

blumamir replied:
Fixed. I hope I implemented it in a go idiomatic manner (I am new to golang).

If you have the time to take another look that would be amazing

MovieStoreGuy reviewed (changes requested):
There is a lot within this one PR and I am finding it hard to rationalise it all. It also goes against our contribution guides of splitting up an initial component into three PRs.

https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#when-adding-a-new-component

blumamir commented:
> There is a lot within this one PR and I am finding it hard to rationalise it all. It also goes against our contribution guides of splitting up an initial component into three PRs.
> 
> https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#when-adding-a-new-component

Thank you for the thorough review. Should I close this PR and reopen it as 3 PRs as in the link you sent? or continue with it as is?

MovieStoreGuy commented:
> Thank you for the thorough review. Should I close this PR and reopen it as 3 PRs as in the link you sent? or continue with it as is?

Since @atoulme is the sponsor for this PR, I will delegate it this to them if it should be split up or not.

MovieStoreGuy commented:
I'm happy to remove my request changes once:

- [ ] Removal of pointers to `pdata` types
- [ ] Move to using "set" look ups over `slices.Contains`

I am on the fence about `ScopeLeaves` since I worry it is not a term I've seen often when dealing with tracing so having to explain the intent could be a little much.

atoulme commented:
3 PRs is easier. Usually what I do is take from the big PR and apply a smaller portion of the diff into a PR that can be easily groked. The original PR stays open, and rebases off main every time we chip away until it's a simple PR itself.

Commit:
chore: add blumamir as codeowner

Commit:
chore: fix markdown alignment

Commit:
chore: fix comment

Commit:
test: add Parallel in tests

Commit:
chore: update go version

Commit:
fix: make FullSpan type private

Commit:
fix: make TraceTreeData private

Commit:
pref: convert slice contains with map

Commit:
Merge remote-tracking branch 'upstream/main' into in-trace-sampler

Commit:
chore: bump collector versions

Commit:
fix: shadow declaration

Commit:
chore: make gendependabot

MovieStoreGuy reviewed (commented):
Few updates.

Commit:
Merge remote-tracking branch 'upstream/main' into in-trace-sampler

Commit:
chore: add comment about fnvhasher original source and motivation

Commit:
refactor: remove reference to pdata

Commit:
chore: collector version 0.74.0

Commit:
chore: make gendependabot

blumamir commented:
> There is a lot within this one PR and I am finding it hard to rationalise it all. It also goes against our contribution guides of splitting up an initial component into three PRs.

Created the first PR #20319. Hope it is makes things simpler, sorry about that 🙏

@atoulme @MovieStoreGuy

atoulme commented:
Closing this PR in favor of https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/20319.
